scipy/optimize/_direct.py

```diff
@@ -0,0 +1,203 @@
+    callback : callable, `callback(xk)`, optional
+        A function to follow the progress of the minimization. ``xk`` is
+        the best solution found so far.
+
+    Returns
+    -------
+    res : OptimizeResult
+        The optimization result represented as a ``OptimizeResult`` object.
+        Important attributes are: ``x`` the solution array, ``success`` a
+        Boolean flag indicating if the optimizer exited successfully and
+        ``message`` which describes the cause of the termination. See
+        `OptimizeResult` for a description of other attributes.
+
+    Notes
+    -----
+
+    DIRECT is a deterministic
+    optimization algorithm capable of minimizing black box function with
+    its variables subject to lower and upper bound constrains by sampling
+    potential solutions in the search space. The algorithm starts with
+    mapping the hyperrectangle (set of all possible values that can be taken
+    by the input variables subject to the bound constraints) n-dimensional
+    unit hypercube. It samples the function at the center of this hypercube
+    and at 2n (n is the number of variables) more points, 2 in each coordinate
+    direction. Using these function values, DIRECT then divides the domain
+    into hyperrectangles, each having exactly one of the sampling points as
+    its center. In each iteration, DIRECT chooses, using the epsilon parameter,
+    by default 1e-4, some of the existing hyperrectangles to be further divided
+    . This division process continues until the maximum iterations or maximum
+    function evaluations allowed are exceeded, or the function value is within
+    the desired percentage error of the global minimum (if known). The improved
+    version of DIRECT algorithm [1]_ is biased towards local search making it
+    effective for functions without too many local minima.
+
+    This code is based on the DIRECT 2.0.4 Fortran code by Gablonsky et al. at
+    http://www4.ncsu.edu/~ctk/SOFTWARE/DIRECTv204.tar.gz
+    The C version was initially converted via f2c and then cleaned up and
+    reorganized by Steven G. Johnson, August 2007. And this method wraps
+    the C implementation.
+
+    .. versionadded:: 1.8.0
+
+    References
+    ----------
+    .. [1] Jones, D.R., Perttunen, C.D. & Stuckman, B.E. Lipschitzian
+           optimization without the Lipschitz constant. J Optim Theory Appl
+           79, 157-181 (1993)
+    .. [2] Jorg Maximilian Xaver Gablonsky and Carl Timothy Kelley. 2001.
+           Modifications of the direct algorithm. Ph.D. Dissertation.
+    """
+    if not isinstance(bounds, Bounds):
+        lb, ub = old_bound_to_new(bounds)
+        bounds = Bounds(lb, ub)
+
+    lb = np.ascontiguousarray(bounds.lb)
+    ub = np.ascontiguousarray(bounds.ub)
+
+    def _func_wrap(x, *args):
+        x = np.asarray(x)
+        f = func(x, *args)
+        if np.isnan(f):
+            return np.nan
+        else:
+            return f
```

Comment:
```suggestion
        return f
```
Maybe, I'm missing domething.